### PR TITLE
gpio: add support for get_config and get_direction in the nct38xx driver

### PR DESCRIPTION
**Pull request by barnas-michal**
This PR adds the implementations for the API used to get current configurations of the pins in the NCT38xx IO expanders family.

### CODE REVIEW

File: drivers/gpio/gpio_nct38xx_port.c [hunk 1]
```diff
@@ -6,11 +6,10 @@
 
 #define DT_DRV_COMPAT nuvoton_nct38xx_gpio_port
 
-#include <zephyr/drivers/gpio.h>
```
Comment (cfriedt) [inline drivers/gpio/gpio_nct38xx_port.c:9]: Whitespace changes (including header re-ordering) are best to be kept in a separate commit.

File: drivers/gpio/gpio_nct38xx_port.c [hunk 2]
```diff
@@ -70,7 +69,7 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
 		}
 
 		new_reg = reg | mask;
-		/* NCT3807 bit3 should be set to 0 */
```
Comment (cfriedt) [inline drivers/gpio/gpio_nct38xx_port.c:73]: probably should be a separate commit from the gpio dir + config commit

File: drivers/gpio/gpio_nct38xx_port.c [hunk 3]
```diff
@@ -459,16 +580,16 @@ BUILD_ASSERT(CONFIG_GPIO_NCT38XX_PORT_INIT_PRIORITY > CONFIG_GPIO_NCT38XX_INIT_P
 
 #define GPIO_NCT38XX_PORT_DEVICE_INSTANCE(inst)                                                    \
 	static const struct gpio_nct38xx_port_config gpio_nct38xx_port_cfg_##inst = {              \
-		.common = { .port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(inst) &               \
-					     DT_INST_PROP(inst, pin_mask) },                       \
+		.common = {.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(inst) &                \
```
Comment (cfriedt) [inline drivers/gpio/gpio_nct38xx_port.c:583]: this and the change below are only whitespace. should be in a separate commit

File: drivers/gpio/gpio_nct38xx_port.c (hunk 2 above)
Comment (semihalf-barnas-michal) [reply drivers/gpio/gpio_nct38xx_port.c:73]: It is separate commit, just the same PR

**Review by cfriedt — CHANGES_REQUESTED** [review]
Looks good - just some nits. Please split this into two commits:
1. commit for adding gpio dir and config
2. whitespace, setting `ret`, comment, and header reordering "clean up"

File: drivers/gpio/gpio_nct38xx_port.c [hunk 4]
```diff
@@ -131,8 +130,8 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
 
 	/* Configure pin as output, if requested 0:input 1:output */
 	if (flags & GPIO_OUTPUT) {
-		nct38xx_reg_read_byte(config->nct38xx_dev, NCT38XX_REG_GPIO_DIR(config->gpio_port),
```
[the remark this answers is not in the document]
Comment (semihalf-barnas-michal) [reply drivers/gpio/gpio_nct38xx_port.c:134]: It is in separate commit from the gpio config/dir, it's in the same commit as the change of "should -> must"

File: drivers/gpio/gpio_nct38xx_port.c (hunk 3 above)
Comment (semihalf-barnas-michal) [reply drivers/gpio/gpio_nct38xx_port.c:583]: Done

**Review by cfriedt — DISMISSED** [review]

**Comment by cfriedt** [comment]
@semihalf-barnas-michal - can you please rebase. I'll approve again. 

@henrikbrixandersen - PTAL

File: drivers/gpio/gpio_nct38xx_port.c [hunk 5]
```diff
@@ -146,6 +146,78 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
...
+	const struct gpio_nct38xx_port_config *const config = dev->config;
+	struct gpio_nct38xx_port_data *const data = dev->data;
+	uint32_t mask = BIT(pin);
+	uint8_t reg;
+	int ret;
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (config->gpio_port == 0) {
+		if (mask & (~config->common.port_pin_mask)) {
```
Comment (sjg20) [inline drivers/gpio/gpio_nct38xx_port.c:160]: Please drop the extra brackets as they are confusing

File: drivers/gpio/gpio_nct38xx_port.c [hunk 6]
```diff
@@ -146,6 +146,78 @@ static int gpio_nct38xx_pin_config(const struct device *dev, gpio_pin_t pin, gpi
...
+			ret = -ENOTSUP;
+			goto done;
+		}
+
+		ret = nct38xx_reg_read_byte(config->nct38xx_dev, NCT38XX_REG_MUX_CONTROL, &reg);
+		if (ret < 0) {
+			goto done;
+		}
+
+		if ((mask & config->pinmux_mask) && (mask & (~reg))) {
```
Comment (sjg20) [inline drivers/gpio/gpio_nct38xx_port.c:170]: Can simplify to:

mask & config->pinmux_mask & ~reg

How about adding a comment too?

File: drivers/gpio/gpio_nct38xx_port.c [hunk 7]
```diff
@@ -381,6 +453,50 @@ static int gpio_nct38xx_manage_callback(const struct device *dev, struct gpio_ca
...
+
+	/* Read direction register, 0 - input, 1 - output */
+	ret = nct38xx_reg_read_byte(config->nct38xx_dev, NCT38XX_REG_GPIO_DIR(config->gpio_port),
+				    &dir_reg);
+	if (ret < 0) {
+		goto done;
+	}
+
+	if (inputs) {
+		*inputs = mask & (~dir_reg);
```
Comment (sjg20) [inline drivers/gpio/gpio_nct38xx_port.c:486]: same here - drop brackets

**Review by sjg20 — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]
